prov/verbs/src/fi_verbs.h

```diff
@@ -364,11 +364,15 @@ struct vrb_domain {
 		int			xrcd_fd;
 		struct ibv_xrcd		*xrcd;
 
-		/* The domain maintains a RBTree for mapping an endpoint
-		 * destination addresses to physical XRC INI QP connected
-		 * to that host. The map is protected using the EQ lock
-		 * bound to the domain to avoid the need for additional
-		 * locking. */
+		/* XRC INI QP connections can be shared between endpoint
+		 * within the same domain. The domain maintains an RBTree
+		 * for mapping endpoint destination addresses to the
+		 * physical XRC INI connection to the associated node. The
+		 * map and XRC INI connection object state information are
+		 * protected via the ini_lock. */
+		fastlock_t		ini_lock;
+		ofi_fastlock_acquire_t	lock_acquire;
+		ofi_fastlock_release_t	lock_release;
```

Comment:
Is this lock ever acquired as part of any fast path operation?  If not, can we drop the optimization and always just use it?

Reply:
No it is never acquired as part of the fast path, but at very large scale there are a ton of connections being done and we still won't those to be responsive. Keeping it means MPI with FI_THREAD_DOMAIN can avoid taking the lock since the EQ lock offers sufficient protection. With that said, if you feel strongly that it is unnecessary complexity than I can remove if @jswaro is ok with that as well.

Reply:
I'm okay leaving the handlers.

Reply:
@shefty - Also, any follow-on fine grain locking is going to want to take advantage of the acquire/release NOOP functions as well.

Reply:
The noops prevent us from taking the lock when we don't need to. Additionaly, like Steve mentioned, he's working on a finer grained approach that addresses lock contention problems when there are multiple threads in a rank. The finer-grained approach that he's working on is intended to address a potential performance issue we expect with MPI and Chapel where they could run a single rank with upwards of 64 threads -- instead of 64 ranks with a 1-3 threads.

Reply:
@shefty - Thanks.